firedrake/slate/slate.py

```diff
@@ -296,6 +305,107 @@ def _key(self):
         return (type(self), self._function)
 
 
+class IndexedTensor(TensorBase):
+    """This class represents a tensor corresponding
+    to particular block of a mixed tensor.
+
+    :arg tensor: A block (mixed) tensor.
+    :arg idx: A tuple of indices denoting which block
+              to extract.
+    """
+
+    def __init__(self, tensor, idx):
+        """Constructor for the IndexedTensor class."""
+        if not isinstance(tensor, TensorBase):
+            raise TypeError("Can only index Slate expressions.")
+
+        if tensor.rank == 0:
+            raise ValueError("Cannot index scalars.")
+
+        assert len(idx) == tensor.rank, (
+            "Number of indices must match the tensor rank."
+        )
```

Comment:
This case also catches indexing a "scalar" with more than zero indices.  So again, I think it makes sense to allow `IndexedTensor(scalar, ())`